[16.0][IMP]fastapi: enable multi-slash routes

PicchiSeba opened the pull request:
This is a Work In Progress PR, which attempts to address this issue https://github.com/OCA/rest-framework/issues/446.
Any suggestion is welcome

EDIT: depends on https://github.com/OCA/rest-framework/pull/524

PicchiSeba committed:
[IMP]fastapi: obtain endpoint via common method

PicchiSeba commented:
The error seems unrelated to this PR

```
AttributeError: module 'marshmallow' has no attribute 'pprint'
```

PicchiSeba committed:
[IMP]fastapi: enable multi-slash routes

lmignon commented on fastapi/models/fastapi_endpoint.py:
```diff
@@ -210,9 +210,24 @@ def _reset_app_cache_marker(self):
         """
 
+    @api.model
+    def get_endpoint(self, root_path):
```
The result should be cached to avoid a lot of useless process on each query???

lmignon reviewed (changes requested):
Than you for the proposal @PicchiSeba

`get_uid` , `get_app` and `get_endpoint` no more receive a root path. They receive a path... The parameter name should be renamed. (see also my previous comment)

lmignon committed:
[IMP] fastapi: Ensures endpoint matching is based on path parts

When we lookup the fastapi.endpoint to use to process a request at a given path, we must ensure that the matching mechanism on the path and the endpoint root_path is bases on the path parts. e.g. /a/b is not prefix of /a/bc.
In the same time improves robustness of the matching mechanism and avoid useless SQL queries

lmignon commented on fastapi/models/fastapi_endpoint.py:
```diff
@@ -211,8 +211,24 @@ def _reset_app_cache_marker(self):
     @api.model
-    def get_app(self, root_path):
-        record = self.search([("root_path", "=", root_path)])
+    @tools.ormcache("path")
+    def get_endpoint(self, path):
+        # try to match the request url with the most similar endpoint
+        endpoints_by_length = self.search([]).sorted(
+            lambda fe: len(fe.root_path), reverse=True
+        )
+        endpoint = False
+        while endpoints_by_length:
+            candidate_endpoint = endpoints_by_length[0]
+            if path.startswith(candidate_endpoint.root_path):
+                endpoint = candidate_endpoint
+                break
+            endpoints_by_length -= candidate_endpoint
+        return endpoint
```
This method should handle the path as a path :smirk:. The use of the `startswith`method doesn't mean that your endpoint path is a sub url of the given path...  e.g. /a/b is not prefix of /a/bc.

see https://github.com/PyTech-SRL/rest-framework/pull/1

lmignon reviewed (changes requested):
see https://github.com/PyTech-SRL/rest-framework/pull/1

lmignon commented:
once https://github.com/OCA/rest-framework/pull/527 is merged, a rebase will solve the CI

lmignon reviewed (approved):
Thank you for this great improvement @PicchiSeba 

LGTM (Code review, functionnal tests)

PicchiSeba commented:
> Thank you for this great improvement @PicchiSeba
> 
> LGTM (Code review, functionnal tests)

Thank you for your support and technical expertize!

lmignon commented:
ping @AnizR @sbidoul

AnizR commented on fastapi/tests/test_fastapi_multislash_demo.py:
```diff
@@ -0,0 +1,32 @@
+    """
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        cls.default_fastapi_router = demo_router
+        cls.default_fastapi_running_user = cls.env.ref(
+            "fastapi.fastapi_endpoint_multislash_demo"
+        )
+        cls.default_fastapi_authenticated_partner = cls.env["res.partner"].create(
+            {"name": "FastAPI Demo"}
+        )
+
+    def test_hello_world(self) -> None:
+        with self._create_test_client() as test_client:
+            response: Response = test_client.get("/demo/")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertDictEqual(response.json(), {"Hello": "World"})
```
I think that this test class should be re-worked:
You are not using your 'fastapi_endpoint_multislash_demo' correctly because you tried to put it on the 'default_fastapi_running_user' (defines the user not the endpoint).

PicchiSeba replied:
Hello @AnizR, thanks for you review. I kinda forgot about testing properly with the Odoo test suite.

How would you test it?

I tried playing with the parameters of the `_create_test_client` method in  `common.py`, but I think it just creates a temporary client with the FastAPI app endpoints. I want to test that Odoo correctly mounts the FastAPI app with the specified url.

I also tried creating a test class inheriting from odoo.tests.HttpCase in order to use the `url_open` method but after activating the `fastapi.endpoint` it seems to be unreachable

simahawk replied:
Have you checked that if the registry is in sync? You can call `_handle_registry_sync` directly if not.

AnizR reviewed (commented):
Besides the test class, the code seems good.
Thanks for your contribution!

PicchiSeba committed:
[IMP]fastapi: add test for multi-slash scenario

PicchiSeba commented:
I accidentally squashed the new test into the merge commit, it should be good now

lmignon commented on fastapi/tests/test_fastapi_multislash_demo.py:
```diff
@@ -0,0 +1,32 @@
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/LGPL).
```
IMO this test class is useless and should be removed. Your multi slash route is tested into the `test_multi_slash` method pf the existing test class.

lmignon reviewed (changes requested):
Thank you for the last changes. One little remaining comment.

PicchiSeba commented:
@lmignon I removed the useless test class

lmignon commented:
/ocabot merge minor
